src/atomic.rs

```diff
@@ -242,10 +266,11 @@ impl<T> Atomic<T> {
         ord: Ordering,
         _: &'scope Scope,
     ) -> Result<(), Ptr<'scope, T>> {
-        let fail_ord = strongest_failure_ordering(ord);
-        match self.data.compare_exchange(current.data, new.data, ord, fail_ord) {
-            Ok(_) => Ok(()),
-            Err(previous) => Err(Ptr::from_data(previous)),
+        let previous = self.data.compare_and_swap(current.data, new.data, ord);
+        if previous == current.data {
```

Comment:
Because of the extra comparison, this change adds an extra `cmp` on x86, see https://godbolt.org/g/5uo8AB . Not that it actually matters (I benchmarked it and there's no difference), just FYI.